mutiny-core/src/auth.rs

```diff
@@ -0,0 +1,169 @@
+use crate::error::MutinyError;
+use crate::indexed_db::MutinyStorage;
+use bitcoin::hashes::{sha256, Hash, HashEngine, Hmac, HmacEngine};
+use bitcoin::secp256k1::{ecdsa, All, Message, PublicKey, Secp256k1, SecretKey};
+use bitcoin::util::bip32::{DerivationPath, ExtendedPrivKey};
+use serde::{Deserialize, Serialize};
+use std::str::FromStr;
+use url::Url;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct AuthProfile {
+    pub index: u32,
+    pub name: String,
+    pub used_services: Vec<String>,
+}
+
+impl AuthProfile {
+    pub fn new(index: u32, name: String) -> Self {
+        Self {
+            index,
+            name,
+            used_services: vec![],
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct SigningProfile {
+    pub profile: AuthProfile,
+    hashing_key: SecretKey,
+}
+
+impl SigningProfile {
+    pub fn new(profile: AuthProfile, hashing_key: SecretKey) -> Self {
+        Self {
+            profile,
+            hashing_key,
+        }
+    }
+
+    pub(crate) fn get_secret_key(
+        &self,
+        context: &Secp256k1<All>,
+        xprivkey: ExtendedPrivKey,
+        url: Url,
+    ) -> Result<SecretKey, MutinyError> {
+        // There exists a private hashingKey which is derived by user LN WALLET using m/138'/0 path.
+        let mut engine = HmacEngine::<sha256::Hash>::new(&self.hashing_key.secret_bytes());
+
+        // LN SERVICE full domain name is extracted from login LNURL
+        let host = url.host().ok_or(anyhow::anyhow!("No host"))?;
+
+        // and then hashed using hmacSha256(hashingKey, full service domain name)
+        engine.input(host.to_string().as_bytes());
+        let derivation_mat = Hmac::<sha256::Hash>::from_engine(engine).into_inner();
+
+        // First 16 bytes are taken from resulting hash and then turned into a sequence of 4 u32 values
+        let uints: [u32; 4] = (0..4)
+            .map(|i| u32::from_be_bytes(derivation_mat[(i * 4)..((i + 1) * 4)].try_into().unwrap()))
+            .collect::<Vec<u32>>()
+            .try_into()
+            .expect("slice with incorrect length");
+
+        // which are in turn used to derive a service-specific linkingKey using m/138'/<long1>/<long2>/<long3>/<long4> path
+        let path = DerivationPath::from_str(&format!(
+            "m/138'/{}/{}/{}/{}",
+            uints[0], uints[1], uints[2], uints[3]
+        ))?;
+
+        let key = xprivkey.derive_priv(context, &path)?;
+        Ok(key.private_key)
+    }
+}
+
+#[derive(Clone)]
+pub struct AuthManager {
+    profiles: Vec<SigningProfile>,
```

Comment:
I really don't like all the index access going on with the `Vec` here. Hashmap or something instead?

Reply:
Added some verification